x-pack/libbeat/reader/parquet/parquet_test.go

```diff
@@ -0,0 +1,286 @@
+		t.Run(name, func(t *testing.T) {
+			tempDir := t.TempDir()
+			fName := fmt.Sprintf("%s/%s_%d.parquet", tempDir, "test", i)
+			data := createRandomParquet(t, fName, tc.columns, tc.rows)
+			file, err := os.Open(fName)
+			if err != nil {
+				t.Fatalf("Failed to open parquet test file: %v", err)
+			}
+			defer file.Close()
+
+			cfg := &Config{
+				// we set ProcessParallel to true as this always has the best performance
+				ProcessParallel: true,
+				// batch size is set to 1 because we need to compare individual records one by one
+				BatchSize: 1,
+			}
+			rows := readAndValidateParquetFile(t, cfg, file, data)
+			// asserts of number of rows read is the same as the number of rows written
+			assert.Equal(t, rows, tc.rows)
+		})
+	}
+}
+
+// readAndValidateParquetFile reads the parquet file and validates the data
+func readAndValidateParquetFile(t *testing.T, cfg *Config, file *os.File, data map[string]bool) int {
+	sReader, err := NewBufferedReader(file, cfg)
+	if err != nil {
+		t.Fatalf("failed to init stream reader: %v", err)
+	}
+
+	rowCount := 0
+	for sReader.Next() {
+		val, err := sReader.Record()
+		if err != nil {
+			t.Fatalf("failed to read stream: %v", err)
+		}
+		if val != nil {
+			rowCount++
+			// this is where we check if the column values are the same as the ones we wrote
+			if !data[string(val)] {
+				t.Fatalf("failed to find record in parquet file: %v", err)
+			}
+		}
+	}
+	return rowCount
+}
+
+// createRandomParquet creates a parquet file with random data
+func createRandomParquet(t testing.TB, fname string, numCols int, numRows int) map[string]bool {
+	// defines a map to store the parquet data for validation
+	data := make(map[string]bool)
+	// creates a new Arrow schema
+	var fields []arrow.Field
+	for i := 0; i < numCols; i++ {
+		fieldType := arrow.PrimitiveTypes.Int32
+		field := arrow.Field{Name: fmt.Sprintf("col%d", i), Type: fieldType, Nullable: true}
+		fields = append(fields, field)
+	}
+	schema := arrow.NewSchema(fields, nil)
+	file, err := os.Create(fname)
+	if err != nil {
+		t.Fatalf("Failed to create parquet test file: %v", err)
+	}
+
+	// creates a new file writer
+	fileWriter, err := pqarrow.NewFileWriter(schema, file, nil, pqarrow.ArrowWriterProperties{})
+	if err != nil {
+		t.Fatalf("Failed to create parquet file writer: %v", err)
+	}
+
+	// creates an Arrow memory pool for managing memory allocations
+	memoryPool := memory.NewGoAllocator()
+
+	// uses a fixed seed value of 1 for generating random data
+	seed := int64(1)
+	r := rand.New(rand.NewSource(seed))
+
+	// generates random data for writing to the parquet file
+	for rowIdx := int64(0); rowIdx < int64(numRows); rowIdx++ {
+		// creates an Arrow record with random data
+		var recordColumns []arrow.Array
+		for colIdx := 0; colIdx < numCols; colIdx++ {
+			randData := []int32{r.Int31()}
+			builder := array.NewInt32Builder(memoryPool)
+			builder.AppendValues(randData, nil)
+			defer builder.Release()
```

Comment:
This function also has `defer` in a loop in two places. It should be able to release the resources as soon as possible  instead of waiting until all rows/columns are generated.